### PR TITLE
Performance improvements for restic-exporter

### PR DESCRIPTION
**Pull request by k0ste**
Hi, here is some fight with `restic` cloud-native repo variant (S3)
With options, that disables the `restic check` & `restic stats` queries - the duration of metrics collecting reduces grammatically 
```python
restic_scrape_duration_seconds 137.99019384384155
restic_scrape_duration_seconds 51.43253993988037
```
`-e NO_CHECK="true" -e NO_STATS="true"`: 
```python
restic_scrape_duration_seconds 34.870311975479126
restic_scrape_duration_seconds 16.271262645721436
```
For "backup is not performed for a time" - the timestamp + meta metrics just enough 

Full list of changes:
* Added `EXIT_ON_ERROR` env var - now is possible to control "exit_on_errors"
behaviour
* Added `NO_CHECK` env var, now is possible not perform `restic check` operation
* Added `NO_STATS` env var, now is possible not collect per backup stats
* Added backup `tag` to the backup label (if `tags` is present)
* Code formatted via `black` formatter

### CODE REVIEW

File: restic-exporter.py [hunk 1]
```diff
@@ -10,149 +10,228 @@
...
+                stats = {
+                    "total_size": -1,
+                    "total_file_count": -1,
+                }
+            else:
+                stats = self.get_stats(snap["id"])
+
+            # use first element of tags if tags is present
+            if "tags" in snap:
+                tag = snap["tags"][0]
```
Comment (ngosang) [inline restic-exporter.py:157]: Why not add all tags? Prometheus query language allows regular expressions.

File: README.md [hunk 2]
```diff
@@ -97,38 +97,59 @@ All configuration is done with environment variables:
...
 # TYPE restic_backup_timestamp gauge
-restic_backup_timestamp{client_hostname="PC-HOME-1",client_username="PC-HOME-1\\User-1",snapshot_hash="1911eb846f1642c327936915f1fad4e16190d0ab6b68e045294f5f0280a00ebe"} 1.669754009e+09
+restic_backup_timestamp{snapshot_hash="20795072cba0953bcdbe52e9cf9d75e5726042f5bbf2584bb2999372398ee835"} 1.666273638e+09
+# HELP restic_backup_meta Metadata of the backup
+# TYPE restic_backup_meta gauge
+restic_backup_meta{client_hostname="product.example.com",client_username="root",snapshot_hash="20795072cba0953bcdbe52e9cf9d75e5726042f5bbf2584bb2999372398ee835",snapshot_tag="mysql"} 1.0
 # HELP restic_backup_files_total Number of files in the backup
 # TYPE restic_backup_files_total counter
-restic_backup_files_total{client_hostname="PC-HOME-1",client_username="PC-HOME-1\\User-1",snapshot_hash="1911eb846f1642c327936915f1fad4e16190d0ab6b68e045294f5f0280a00ebe"} 19051.0
+restic_backup_files_total{snapshot_hash="20795072cba0953bcdbe52e9cf9d75e5726042f5bbf2584bb2999372398ee835"} 8.0
```
Comment (ngosang) [inline README.md:143]: Please, revert this change. I want all the labels in all the metrics.
1. You are breaking backward compatibility and the provided Grafana Dashboard.
2. They don't use any space in Prometheus. All the labels are unchanged for each client so Prometheus only creates 1 timeserie and it stores the values.
3. The queries will be more complex to draw the client info together with the metric.

**Review by ngosang — COMMENTED** [review]
Thank you for the changes. I let you some comments. If you make the changes I will publish a release this weekend.

**Review by ngosang — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -97,38 +97,59 @@ All configuration is done with environment variables:
...
+default is `8001`.
+- `LISTEN_ADDRESS`: (Optional) The address the exporter should listen on. The
+default is to listen on all addresses.
 - `LOG_LEVEL`: (Optional) Log level of the traces. The default is `INFO`.
+- `ANY_ERROR_FATAL`: (Optional) Shutdown exporter on any `restic` error. Default
+is `Flase` (only log error, such as network error with Cloud backends).
+- `NO_CHECK`: (Optional) Do not perform `restic check` operation for performance
+reasons. Default is `False` (perform `restic check`).
+- `NO_STATS`: (Optional) Do not collect per backup statistics for performance
+reasons. Default is False (collect per backup statistics).
```
[the remark this answers is not in the document]
Comment (k0ste) [reply README.md:124]: Thanks

File: restic-exporter.py [hunk 4]
```diff
@@ -185,62 +269,86 @@ def get_stats(self, snapshot_id=None):
...
         restic_repo_password_file = os.environ["RESTIC_REPO_PASSWORD_FILE"]
     except Exception:
-        logging.error("Configuration error. The environment variable RESTIC_REPO_PASSWORD_FILE is mandatory")
+        logging.error("The environment variable RESTIC_REPO_PASSWORD_FILE is mandatory")
         sys.exit(1)
 
     exporter_address = os.environ.get("LISTEN_ADDRESS", "0.0.0.0")
     exporter_port = int(os.environ.get("LISTEN_PORT", 8001))
     exporter_refresh_interval = int(os.environ.get("REFRESH_INTERVAL", 60))
+    exporter_any_error_fatal = bool(os.environ.get("ANY_ERROR_FATAL", False))
```
[the remark this answers is not in the document]
Comment (k0ste) [reply restic-exporter.py]: Definitely. On any connection error, for example:

* link is down
* arp problems on switch
* L3 drops
* S3 endpoint unreachable

The application will be exit with error code, and docker should restart the whole app. Instead this - better to wait when something (like network ops finish) just fixes, or admin comes to read exporter log's to determine what is actually wrong

File: restic-exporter.py (hunk 1 above)
Comment (k0ste) [reply restic-exporter.py:157]: Cause what to do with tags next? We could make `k,v` tags, for example:

```python
__name__{tag1="value1", tag2="value2"}
```

but tags is a one `list[]`, so all tags will be:

```
__name__{tags="tag1,tag2"}
```

What to do with this tags next, with alert? 
What do we do with tags?

* put e-mail address to tags - to this e-mail address alerts will be pushed (the actual owner of the backup)
* put application name to tags - send alerts to chat, where actual owner catch hist application
* use alertmanager routing for finding the receiver by app name/team name/project

**Comment by k0ste** [comment]
> Thank you for the changes. I let you some comments. If you make the changes I will publish a release this weekend.

I was made changes and reverts, as you suggested 🙏

**Review by ngosang — APPROVED** [review]
Thank you!

**Comment by ngosang** [comment]
https://github.com/ngosang/restic-exporter/releases/tag/1.2.0
